### ISSUE TITLE
Bug: no-useless-escape rule fails to identify valid regex escape.

### ISSUE THREAD

**Issue by raibtoffoletto** [eslint/eslint#15044]
### Environment

Node version: v15.12.0
npm version: v7.6.3
Local ESLint version: v7.27.0
Global ESLint version: Not found
Operating System: linux 5.4.0-81-generic

### What parser are you using?

@typescript-eslint/parser

### What did you do?

```markdown
<details>
<summary>Configuration</summary>
Rule fails to identify the necessary escape in RegExp constructor.


new RegExp(`(\d#|)`) // fails
new RegExp('(\d#|)') // fails
/(\d#|)/ // pass

</details>
```

### What did you expect to happen?

```markdown
Expected to recognized that the escape on `\d` is necessary to distinguish between the digit character identifier: **\d**, and the character `D` identifier: **d**.
```

### What actually happened?

```markdown
A warning saying the the expression `new RegExp('(\d#|)')` has an useless escape, suggesting to use **d** in place of **\d**. This difference leads to completely different regex parsing, thus making the escape **\** NOT useless.


Unnecessary escape character: \d. [eslint no-useless-escape]
```
```


### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by mdjermanovic** [comment]
Hi @raibtoffoletto, thanks for the issue!

This works as expected, `\d` is a useless escape in string literals and template literals, and it evaluates to `'d'` character.

```js
'(\d#|)' === '(d#|)'; // true
```

Consequently, `new RegExp('(\d#|)')` creates a regular expression with pattern `(d#|)`, not `(\d#|)`.

```js
const regex = new RegExp('(\d#|)');

regex.toString(); // '/(d#|)/'

regex.exec('1#'); // ['', '', index: 0, input: '1#', groups: undefined]

regex.exec('d#'); // ['d#', 'd#', index: 0, input: 'd#', groups: undefined]
```

**Comment by raibtoffoletto** [comment]
Dear @mdjermanovic,
Thanks for the answer and explanation. I'm using a template string to construct the regex and it does match digits characters... and also a 'D' char. So I think it is somehow a problem with JS Regex implementation (?).

The logic I'm using is this...
https://codesandbox.io/s/regextest-uvhue?file=/index.js

In any case es-lint is write,
and sorry for opening an unnecessary issue.

**Comment by mdjermanovic** [comment]
> I'm using a template string to construct the regex and it does match digits characters... and also a 'D' char

Unescaped `d` matches only `d` character, but in `(d#|)` you have a zero-length pattern in an alternative.

When you run `/(d#|)col1/` on `2#col1`, `(d#|)` matches with a zero-length string and `col1` with `"col1"`.

```js
/(d#|)col1/.exec("2#col1"); // ['col1', '', index: 2, input: '2#col1', groups: undefined]
```

If you want a regular expression that would test if the string is either `id` or `<digit>#id`, this would do the work:

```js
new RegExp(`^(\\d#)?${idToFind}$`, "i");
```
